core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -210,15 +210,24 @@ private[chisel3] trait HasId extends InstanceId {
   // Uses a namespace to convert suggestion into a true name
   // Will not do any naming if the reference already assigned.
   // (e.g. tried to suggest a name to part of a Record)
-  private[chisel3] def forceName(prefix: Option[String], default: => String, namespace: Namespace): Unit =
+  private[chisel3] def forceName(
+    prefix:     Option[String],
+    default:    => String,
+    namespace:  Namespace,
+    errorIfDup: Boolean = false,
```

Comment:
```suggestion
    // Resolve name collisions, typically uses namespace.name
    resolveName:  String => String
```
This logic feels pretty convoluted to me but we can fold the two arguments into one which I think makes the erroring more clear. The exception throwing can then move out of this function (which feels wrong) and into the caller site when naming a port.